doc/development/autograph.rst

```diff
@@ -0,0 +1,743 @@
+[Array(-0.45165857, dtype=float64)]
+
+Currently, AutoGraph supports converting the following Python statements:
+
+- ``if`` statements (including ``elif`` and ``else``)
+- ``for`` loops
+- ``while`` loops
+
+``break`` and ``continue`` statements are currently not supported. The logical operators
+``and``, ``or`` and ``not`` are currently unsupported.
+
+Nested functions
+----------------
+
+AutoGraph will continue to work even when the function
+itself calls nested functions. All functions called within the
+top-level function will also have Python control flow captured
+and converted by AutoGraph.
+
+In addition, built-in functions from ``jax``, ``pennylane``, and ``catalyst``
+are automatically *excluded* from the AutoGraph conversion.
+
+.. code-block:: python
+
+    def f(x):
+        if x > 5:
+            y = x ** 2
+        else:
+            y = x ** 3
+        return y
+
+    def g(x, n):
+        for i in range(n):
+            x = x + f(x)
+        return x
+
+>>> plxpr = make_plxpr(g)(0.0, 1)  # initialize with arguments of correct type and shape
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, 0.4, 6)
+[Array(22.14135448, dtype=float64)]
+
+If statements
+-------------
+
+While most ``if`` statements you may write in Python will be automatically
+converted, there are some important constraints and restrictions to be aware of.
+
+Return statements
+~~~~~~~~~~~~~~~~~
+
+Return statements are generally supported inside of ``if``/``elif``/``else`` statements,
+however, the returned values require a matching shape and structure across branches.
+
+For example, consider the following pattern, where two different array dimensions are returned
+from each branch:
+
+.. code-block:: python
+
+    def f(x):
+        if x > 5:
+            return jnp.array([1, 2])
+        return jnp.array([0])
+
+This will generate the following error:
+
+>>> make_plxpr(f)(0)
+ValueError: Mismatch in output abstract values in false branch #0 at position 1:
+ShapedArray(int64[1]) vs ShapedArray(int64[2])
+
+This is relevant for any example that uses different *structure* across branches. The structure of a function
+output is defined by things like the number of results, the containers used like lists or
+dictionaries, or more generally any (compile-time) PyTree metadata.
+
+Different branches must assign the same type
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Different branches of an if statement must always assign variables with the same type across branches,
+if those variables are used in the outer scope (external variables). The type must be the same in the sense
+that the *structure* of the variable should not change across branches, and the dtypes must match.
+
+In particular, this requires that if an external variable is assigned an array in one
+branch, other branches must also assign arrays of the same shape. Consider this function, which has the
+same return shape regardless of branch, but differs in the shape of `y` in difference logic branches:
```

Comment:
Not just shape, they have to match in shape in shape and type. And for dictionaries they need to match in their nested structure, which I guess is technically not a shape? 

But this also fails, even though the shapes match, because of the mismatch between float and integer:

```
>>> def f(x):
...     if x > 1:
...         y = jnp.array([0.1, 0.2])
...     else:
...         y = jnp.array([1, 2])
...     return jnp.sum(y)
>>> make_plxpr(f)(0.5)
```

I'll update the text and example, they focus a lot of the shape aspect specifically, so I think this doesn't come across as clearly as it could.